lib/agent.js

```diff
@@ -313,153 +314,161 @@ Agent.prototype.addMetadataFilter = function (fn) {
   this._transport.addMetadataFilter(fn)
 }
 
+const EMPTY_OPTS = {}
+
+// Capture an APM server "error" event for the given `err` and send it to APM
+// server.
+//
+// Usage:
+//    captureError(err, opts, cb)
+//    captureError(err, cb)
+//
+// where:
+// - `err` is an Error instance, or a string message, or a "parameterized string
+//   message" object, e.g.:
+//      {
+//        message: "this is my message template: %d %s"},
+//        params: [ 42, "another param" ]
+//      }
+// - `opts` can include any of the following (all optional):
+//   - `opts.timestamp` - Milliseconds since the Unix epoch. Defaults to now.
+//   - `opts.user` - Object to add to `error.context.user`.
+//   - `opts.tags` - Deprecated, use `opts.labels`. Object to add to
+//     `error.context.labels`.
+//   - `opts.labels` - Object to add to `error.context.labels`.
+//   - `opts.custom` - Object to add to `error.context.custom`.
+//   - `opts.message` - If `err` is an Error instance, this string is added to
+//     `error.log.message` (unless it matches err.message).
+//   - `opts.request` - HTTP request (node `IncomingMessage` instance) to use
+//     for `error.context.request`.
+//   - `opts.response` - HTTP response (node `ServerResponse` instance) to use
+//     for `error.context.response`.
+//   - `opts.handled` - Boolean indicating if this exception was handled by
+//     application code.
+//   - `opts.captureAttributes` - Boolean. Default true. Set to false to *not*
+//     include properties of `err` as attributes on the APM error event.
+//   - `opts.skipOutcome` - Boolean. Default false. Set to true to not have
+//     this captured error set `<currentSpan>.outcome = failure`.
+// - `cb` is a callback `function (captureErr, apmErrorIdString)`
 Agent.prototype.captureError = function (err, opts, cb) {
-  if (typeof opts === 'function') return this.captureError(err, null, opts)
+  if (typeof opts === 'function') {
+    cb = opts
+    opts = EMPTY_OPTS
+  } else if (!opts) {
+    opts = EMPTY_OPTS
+  }
 
-  var agent = this
-  var trans = this.currentTransaction
-  var span = this.currentSpan
-  var timestamp = normalizeTimestamp(opts && opts.timestamp)
-  var context = (span || trans || {})._context || { traceparent: {} }
-  var req = opts && opts.request instanceof IncomingMessage
-    ? opts.request
-    : trans && trans.req
-  var res = opts && opts.response instanceof ServerResponse
-    ? opts.response
-    : trans && trans.res
-  var captureAttributes = opts && opts.captureAttributes
-  const skipOutcome = opts && opts.skipOutcome
-  var _isError = isError(err)
-
-  if ((!opts || opts.handled !== false) &&
+  const agent = this
+  let callSiteLoc = null
+  const errIsError = isError(err)
+  const handled = opts.handled !== false // default true
```

Comment:
REVIEW-NOTE: This changes `handled` to be set `true` in cases where it was sometimes left `undefined` by accident/bug. Previously it was set like this: `handled = !opts || opts.handled`. The result of the latter is that adding unrelated opts, e.g. `captureError(err, { request: res}, cb)` changes `handled` from true to undefined.